### ISSUE TITLE
Harden against malicious/abusive users

### ISSUE THREAD

**Issue by KellerFuchs** [hashbang/hashbang#17]
We need to nail this down before publicizing #!:
- Protect user's privacy
  - [x] Prevent users from observing another user's processes
  - [ ] Prevent users from observing another user's network connections
  - [ ] Prevent users from observing login/logoff times
- Prevent resources abuse
  - [x] Restrict number of processes/threads, to protect against fork bombs (`pam_limits`)
  - [ ] Enforce [disk quota](https://github.com/hashbang/pam_setquota)
  - [x] [Limit RAM and CPU usage](https://github.com/hashbang/shell-etc/pull/183) (esp. useful against Bitcoin miners)
- [ ] Prevent [abusive user creation](https://github.com/hashbang/hashbang/blob/master/abuse/signup.md)
- Harden against hostile users trying to “own” the shell server
  - Using a namespace-based sandbox
    - [x] Provide users with a private `/tmp` (and `/var/tmp`, `/run/lock`, ...)
    - [ ] Use a minimal  `/dev` and per-user `/dev/pts`.
    - [ ] Make the mount namespace readonly, except for select directories:
      - `${HOME}`
      - `/var/tmp`, `/tmp` and `/dev/shm` (which are instantiated per-user)
      - `/run/user/${UID}`
      - `/run/lock` if needed (?)

      Rationale: this mitigates issues like CVE-2017-5180 (local root through arbitrary file overwrite in firejail).
      List compiled based on the output of `find / -type d -writable -print -prune` as an unprivileged user.

    - [ ] Prevent users from accessing `/sys`
      Rationale: `/sys` exposes attack surface to virtually every single kernel module.

  - [ ] Apply a `seccomp` syscall filter to users
    Rationale: this can drastically reduce the attack surface on the kernel.
  - [ ] Deploy a MAC (Mandatory Access Control) system
    Rationale: this helps reduce the attack surface against the system (both kernel and userspace) by constraining users.
  - ~~Use a PaX/GrSec kernel~~
    ~~Rationale: this is a last line of defense against (mostly) kernel vulnerabilities, and makes reliable exploitation much harder.~~
   GrSec has gone closed-source.

- Harden against malicious users trying to exploit other users
    - [x] Harden and extend our default [firejail](https://github.com/hashbang/shell-etc/issues/86) profiles
      Rationale: this makes it harder to “own” users through applications that we start by default (`weechat`, `mutt`, ...)
    - [x] Provide per-user `/tmp` & co through namespacing (see above)
    - [ ] Make `/home` read-only in the mount namespace.

**Comment by KellerFuchs** [comment]
The “harden against hostile users” section contains relatively “wild” ideas, which are there for discussion.

**Comment by daurnimator** [comment]
> Protect user's privacy

I don't think this is a blocker before publicizing. It's just a 'nice to have'.
Usually shells don't even have as much privacy as we already have 

---

>  Prevent users from accessing /sys

Why?

> Use a PaX/GrSec kernel

no thanks

> Deploy a MAC (Mandatory Access Control) system

I'd rather not. it makes the shell servers very unapproachable and unfamiliar.

**Comment by KellerFuchs** [comment]
> > Protect user's privacy
> 
> I don't think this is a blocker before publicizing. It's just a 'nice to have'.
> Usually shells don't even have as much privacy as we already have 

I would argue to the contrary.
Perhaps other shells protect their users less, but we aim to be a beginner-friendly environment, and having other users snooping on your open connections or when you connect would be very off-putting.

> >  Prevent users from accessing /sys
> 
> Why?

To reduce the attack surface against the kernel, since `/sys` provide an interface to virtually all loaded kernel modules.

> > Use a PaX/GrSec kernel
> 
> no thanks

Why?

> > Deploy a MAC (Mandatory Access Control) system
> 
> I'd rather not. it makes the shell servers very unapproachable and unfamiliar.

Being “something the user doesn't have to care about 99.9999% of the time” would definitely be a design goal.  Assuming we set a user-wide policy (so, something that describes their access rights), I don't really see the potential for confusion.

**Comment by KellerFuchs** [comment]
daurnimator and I [discussed](https://github.com/hashbang/hashbang/files/451303/grsec.txt) the case of grsec over IRC, and what came out was that:
- It can be a stability issue, so we would have to keep an eye on this.
- We shouldn't build things that rely on grsec-specific features to be secure.  
  To enforce this, the easiest would be to disable:
  - [Role-Based Access Control](https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Role_Based_Access_Control_Options);
  - [chroot restrictions](https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Chroot_jail_restrictions);
  - [shared-memory restrictions](https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Disallow_access_to_overly-permissive_IPC_objects);
  - [symlinks restrictions](https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Kernel-enforced_SymlinksIfOwnerMatch).

daurnimator said he would be OK with that, though he would rather not have grsec at all.

**Comment by RyanSquared** [comment]
@KellerFuchs can you edit the above comment to also include the reasons we are using GrSec, not ujust the things we're disabling?

**Comment by KellerFuchs** [comment]
@ChickenNuggers The reason for GrSec is in the issue description: hardening the shell server itself against malicious users.

GrSec is a kernel patch that uses a variety of techniques to harden it against exploitation.  It has a pretty good track record when it comes to making the exploitation of a kernel 0day really hard (typically, you end up needing an auxiliary infoleak vulnerability to successfully bypass KASLR and friends).

**Comment by KellerFuchs** [comment]
@hashbang/administrators I updated the wishlist here with rationales and a proposal to make the users' mount namespaces readonly (except for whitelisted locations).

This would mitigate a large number of attack vectors, like today's CVE-2017-5180.

**Comment by daurnimator** [comment]
> Limit RAM and CPU usage (esp. useful against Bitcoin miners)

We should be able to use https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html

**Comment by KellerFuchs** [comment]
@daurnimator Yeah, that was the reason why I opened systemd/systemd#2556

**Comment by daurnimator** [comment]
https://github.com/systemd/systemd/commit/5396624506e155c4bc10c0ee65b939600860ab67 is in systemd now.

**Comment by KellerFuchs** [comment]
@daurnimator Amusingly, systemd 239 is the first release to have that commit, and it's in `stretch-backports`.  O:3
(No, I'm not seriously suggesting to pull `systemd` from backports, that sounds like a PITA)
